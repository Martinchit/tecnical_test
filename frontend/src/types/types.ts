export interface Stock {
  id: number;
  stockId: string;
  currency: string;
  ric: string;
  bloombergTicker: string;
  bloombergTickerLocal: string;
  name: string;
  country: string;
  price: number;
}

export type OrderSide = 'Buy' | 'Sell';
export type Currency = 'HKD' | 'USD' | 'AUD' | 'EUR' | 'SGD';
export type OrderStatus =
  | 'Not Ready'
  | 'Ready'
  | 'In Progress'
  | 'Booked'
  | 'Rejected';
export type OrderExecutionMode = 'Market' | 'Limit';

export interface StockOrder {
  id: number;
  stockId: string;
  currency: string;
  bloombergTickerLocal: string;
  side: OrderSide;
  status: OrderStatus;
  executionMode?: OrderExecutionMode;
  orderPrice?: number;
  shareAmount?: number;
  error?: string;
  selected: boolean;
  orderId: string;
}

export interface StockListAlert {
  [x: number]: OrderSide;
}

export interface ErrorResponse {
  type: string;
  description: string;
  code: number;
}

export interface SelectedId {
  id: number;
  side: string;
}

export type AuthMode = 'Log In' | 'Sign Up';
