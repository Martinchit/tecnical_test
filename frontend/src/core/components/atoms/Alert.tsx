import styled from 'styled-components';
import { Alert as A } from 'react-bootstrap';

export const Alert = styled(A)`
  width: 100%;
  margin: 0;
`;
