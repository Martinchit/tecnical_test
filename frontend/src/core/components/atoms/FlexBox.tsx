import styled from 'styled-components';

export interface FlexBoxProps {
  flexDirection?: string;
  flexWrap?: string;
  justifyContent?: string;
  alignItems?: string;
  width?: string;
  height?: string;
  margin?: string;
  padding?: string;
}

export const FlexBox = styled.div<FlexBoxProps>`
  display: flex;
  flex-direction: ${({ flexDirection }) => flexDirection ?? 'row'};
  flex-wrap: ${({ flexWrap }) => flexWrap ?? 'wrap'};
  justify-content: ${({ justifyContent }) => justifyContent ?? 'center'};
  align-items: ${({ alignItems }) => alignItems ?? 'center'};
  width: ${({ width }) => width ?? '100%'};
  height: ${({ height }) => height ?? '100%'};
  margin: ${({ margin }) => margin ?? 0};
  padding: ${({ padding }) => padding ?? 0};
`;
