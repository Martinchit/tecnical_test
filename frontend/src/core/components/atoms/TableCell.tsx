import styled from 'styled-components';

export const TableCell = styled.td`
  vertical-align: middle !important;
  text-align: center;
`;
