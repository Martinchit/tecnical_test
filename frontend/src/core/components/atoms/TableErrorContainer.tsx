import styled from 'styled-components';

export const TableErrorContainer = styled.td`
  border-top: none !important;
  padding-top: 0 !important;
`;
