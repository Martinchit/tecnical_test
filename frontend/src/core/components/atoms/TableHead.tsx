import styled from 'styled-components';

export const TableHead = styled.th`
  text-align: center;
`;
