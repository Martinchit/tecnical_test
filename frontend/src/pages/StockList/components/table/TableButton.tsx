import styled from 'styled-components';
import { Button } from 'react-bootstrap';

export const TableButton = styled(Button)`
  margin-right: 3%;
  width: 40%;

  @media (max-width: 767px) {
    width: 100%;
    margin-top: 5%;
    font-size: 0.8rem;
  }

  &:focus {
    outline: none;
    box-shadow: none;
  }
`;
