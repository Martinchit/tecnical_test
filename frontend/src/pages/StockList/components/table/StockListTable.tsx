import React from 'react';
import { Table } from 'react-bootstrap';
import { TableButton } from './TableButton';
import { TableErrorContainer } from '../../../../core/components/atoms/TableErrorContainer';
import { TableCell } from '../../../../core/components/atoms/TableCell';
import { TableHead } from '../../../../core/components/atoms/TableHead';
import { Alert } from '../../../../core/components/atoms/Alert';
import { OrderSide, Stock, StockListAlert } from '../../../../types/types';

interface StockListTableProps {
  alert: StockListAlert;
  stocks: Stock[];
  onActionButtonClickHandler: (action: OrderSide, idx: number) => void;
}

export const StockListTable: React.FC<StockListTableProps> = ({
  alert,
  stocks,
  onActionButtonClickHandler,
}) => {
  return (
    <Table responsive>
      <thead>
        <tr>
          <TableHead>Stock Code</TableHead>
          <TableHead>Market Price</TableHead>
          <TableHead>Currency</TableHead>
          <TableHead>Actions</TableHead>
        </tr>
      </thead>
      <tbody>
        {stocks.map(({ id, currency, price, bloombergTickerLocal }, idx) => (
          <React.Fragment key={bloombergTickerLocal}>
            <tr>
              <TableCell>{bloombergTickerLocal}</TableCell>
              <TableCell>{price}</TableCell>
              <TableCell>{currency}</TableCell>
              <TableCell>
                <TableButton
                  variant="primary"
                  onClick={() => onActionButtonClickHandler('Buy', idx)}
                >
                  Buy
                </TableButton>
                <TableButton
                  variant="secondary"
                  onClick={() => onActionButtonClickHandler('Sell', idx)}
                >
                  Sell
                </TableButton>
              </TableCell>
            </tr>
            {alert[id] ? (
              <tr>
                <TableErrorContainer colSpan={4}>
                  <Alert variant="success">
                    {`${alert[id]} order for ${bloombergTickerLocal} is added to the Orders Basket`}
                  </Alert>
                </TableErrorContainer>
              </tr>
            ) : null}
          </React.Fragment>
        ))}
      </tbody>
    </Table>
  );
};
