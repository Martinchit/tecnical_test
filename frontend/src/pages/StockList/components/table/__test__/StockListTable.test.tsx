import React from 'react';
import { shallow } from 'enzyme';

import { StockListTable } from '../StockListTable';
import { TableButton } from '../TableButton';
import { Alert } from '../../../../../core/components/atoms/Alert';
import { OrderSide } from '../../../../../types/types';

describe('StockListTable', () => {
  const props = {
    alert: {},
    stocks: [
      {
        id: 1,
        stockId: 'testOne',
        currency: 'HKD',
        ric: 'testOne',
        bloombergTicker: 'testOne',
        bloombergTickerLocal: 'testOne',
        name: 'testOne',
        country: 'testOne',
        price: 100,
      },
      {
        id: 2,
        stockId: 'testTwo',
        currency: 'HKD',
        ric: 'testTwo',
        bloombergTicker: 'testTwo',
        bloombergTickerLocal: 'testTwo',
        name: 'testTwo',
        country: 'testTwo',
        price: 200,
      },
    ],
    onActionButtonClickHandler: jest.fn(),
  };

  const restoreMock = () => {
    props.onActionButtonClickHandler.mockRestore();
  };

  it('renders correctly', () => {
    const wrapper = shallow(<StockListTable {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('triggers onActionButtonClickHandler on Buy Button click', () => {
    const wrapper = shallow(<StockListTable {...props} />);
    wrapper.find(TableButton).first().simulate('click');
    expect(props.onActionButtonClickHandler).toHaveBeenCalled();
    expect(props.onActionButtonClickHandler).toHaveBeenCalledWith('Buy', 0);
    restoreMock();
  });

  it('triggers onActionButtonClickHandler on Sell Button click', () => {
    const wrapper = shallow(<StockListTable {...props} />);
    wrapper.find(TableButton).at(1).simulate('click');
    expect(props.onActionButtonClickHandler).toHaveBeenCalled();
    expect(props.onActionButtonClickHandler).toHaveBeenCalledWith('Sell', 0);
    restoreMock();
  });

  it('renders alert if value is presented inside the alert props object', () => {
    const side = 'Sell' as OrderSide;
    const alert = {
      1: side,
    };
    const wrapper = shallow(
      <StockListTable
        alert={alert}
        stocks={props.stocks}
        onActionButtonClickHandler={props.onActionButtonClickHandler}
      />
    );
    expect(wrapper.find(Alert).length).toBe(1);
    expect(wrapper.find(Alert).text()).toBe(
      `${side} order for ${props.stocks[0].bloombergTickerLocal} is added to the Orders Basket`
    );
  });
});
