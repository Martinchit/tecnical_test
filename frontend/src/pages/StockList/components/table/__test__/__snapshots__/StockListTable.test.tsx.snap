// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StockListTable renders correctly 1`] = `
<ForwardRef
  responsive={true}
>
  <thead>
    <tr>
      <styled.th>
        Stock Code
      </styled.th>
      <styled.th>
        Market Price
      </styled.th>
      <styled.th>
        Currency
      </styled.th>
      <styled.th>
        Actions
      </styled.th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <styled.td>
        testOne
      </styled.td>
      <styled.td>
        100
      </styled.td>
      <styled.td>
        HKD
      </styled.td>
      <styled.td>
        <Styled(Button)
          onClick={[Function]}
          variant="primary"
        >
          Buy
        </Styled(Button)>
        <Styled(Button)
          onClick={[Function]}
          variant="secondary"
        >
          Sell
        </Styled(Button)>
      </styled.td>
    </tr>
    <tr>
      <styled.td>
        testTwo
      </styled.td>
      <styled.td>
        200
      </styled.td>
      <styled.td>
        HKD
      </styled.td>
      <styled.td>
        <Styled(Button)
          onClick={[Function]}
          variant="primary"
        >
          Buy
        </Styled(Button)>
        <Styled(Button)
          onClick={[Function]}
          variant="secondary"
        >
          Sell
        </Styled(Button)>
      </styled.td>
    </tr>
  </tbody>
</ForwardRef>
`;
